File: src/store/fixtureSource.ts
import { entities, areaRegistry } from "../fixtures/entities";
import { useEntityStore } from "./entityStore";
import { domainOf } from "../lib/ha";
import { slugify, type AutomationConfig } from "../lib/automations";
import { resolveName } from "../lib/resolve";
import { overrides } from "../config/overrides";
import type { HistoryPoint, ServiceData, Source } from "./source";
import type { LogEvent } from "../lib/logbook";
import type { HassEntity } from "../lib/ha";
import {
  DEMO_CLIP_URL,
  DEMO_POSTER_URL,
  type CameraEvent,
} from "../lib/cameraEvents";

// Discrete (on/off-ish) domains step between two states in the synthetic series;
// everything else is treated as a numeric sensor and jittered around its value.
const DISCRETE_DOMAINS = new Set([
  "lock",
  "binary_sensor",
  "switch",
  "light",
  "fan",
  "cover",
  "alarm_control_panel",
  "media_player",
]);

/**
 * Synthesize a plausible history series ending at the entity's current state, so
 * demo mode (no live HA) still renders a chart. Deterministic-ish per entity:
 * numeric sensors wander around their current value; discrete entities flip a
 * few times and settle on the live state.
 */
function synthHistory(entity: HassEntity, hours: number): HistoryPoint[] {
  const now = Date.now();
  const points = Math.min(48, Math.max(12, hours * 2));
  const stepMs = (hours * 3600_000) / points;
  const discrete = DISCRETE_DOMAINS.has(domainOf(entity.entity_id));
  const out: HistoryPoint[] = [];

  if (discrete) {
    const other =
      entity.state === "on"
        ? "off"
        : entity.state === "off"
          ? "on"
          : entity.state === "locked"
            ? "unlocked"
            : "off";
    for (let i = 0; i < points; i++) {
      // Flip on a simple cadence, then guarantee the latest sample is current.
      const flipped = i % 5 === 0 && i !== points - 1;
      out.push({
        t: now - (points - 1 - i) * stepMs,
        state: i === points - 1 ? entity.state : flipped ? other : entity.state,
      });
    }
    return out;
  }

  const base = Number(entity.state);
  const center = Number.isFinite(base) ? base : 50;
  const amp = Math.max(1, Math.abs(center) * 0.08);
  for (let i = 0; i < points; i++) {
    const wobble = Math.sin(i / 3) * amp + (Math.random() - 0.5) * amp * 0.5;
    const value = i === points - 1 ? center : center + wobble;
    out.push({
      t: now - (points - 1 - i) * stepMs,
      state: (Math.round(value * 10) / 10).toString(),
    });
  }
  return out;
}

/** Apply a control action to a fixture entity (simulates HA's state echo). */
function simulate(
  entity: HassEntity,
  domain: string,
  service: string,
  data: ServiceData,
): HassEntity {
  const next: HassEntity = {
    ...entity,
    attributes: { ...entity.attributes },
  };
  if (domain === "lock") {
    next.state = service === "lock" ? "locked" : "unlocked";
  } else if (domain === "light") {
    if (service === "turn_off") {
      next.state = "off";
    } else {
      next.state = "on";
      if (typeof data.brightness_pct === "number") {
        next.attributes.brightness = Math.round((data.brightness_pct / 100) * 255);
      }
    }
  } else if (domain === "switch") {
    next.state = service === "turn_on" ? "on" : "off";
  } else if (domain === "alarm_control_panel") {
    next.state =
      service === "alarm_disarm"
        ? "disarmed"
        : service === "alarm_arm_home"
          ? "armed_home"
          : service === "alarm_arm_away"
            ? "armed_away"
            : entity.state;
  } else if (domain === "cover") {
    if (service === "open_cover") {
      next.state = "open";
      next.attributes.current_position = 100;
    } else if (service === "close_cover") {
      next.state = "closed";
      next.attributes.current_position = 0;
    }
    // stop_cover leaves the resting state as-is (demo has no transit phase).
  } else if (domain === "climate") {
    if (service === "set_temperature" && typeof data.temperature === "number") {
      next.attributes.temperature = data.temperature;
    }
  } else if (domain === "media_player") {
    if (service === "media_play_pause") {
      next.state = entity.state === "playing" ? "paused" : "playing";
    }
    // next/previous track: title is static in demo, state unchanged.
  } else if (domain === "fan") {
    if (service === "turn_off") {
      next.state = "off";
    } else if (service === "turn_on") {
      next.state = "on";
    } else if (service === "set_percentage" && typeof data.percentage === "number") {
      next.state = data.percentage > 0 ? "on" : "off";
      next.attributes.percentage = data.percentage;
    }
  }
  return next;
}

// A plausible message per domain for the synthesized demo logbook.
function demoMessage(entity: HassEntity): string {
  switch (domainOf(entity.entity_id)) {
    case "lock":
      return entity.state === "locked" ? "was locked" : "was unlocked";
    case "binary_sensor":
      return entity.attributes.device_class === "motion"
        ? "detected motion"
        : entity.state === "on"
          ? "was opened"
          : "was closed";
    case "alarm_control_panel":
      return entity.state === "disarmed" ? "was disarmed" : "was armed";
    case "camera":
      return "recorded a clip";
    case "light":
      return entity.state === "on" ? "turned on" : "turned off";
    default:
      return `changed to ${entity.state}`;
  }
}

/**
 * Synthesize a believable home logbook from the current fixtures so demo mode's
 * History hub isn't empty. Events are spread back from `endMs` at a steady
 * cadence and clipped to the requested window.
 */
function synthLogbook(startMs: number, endMs: number): LogEvent[] {
  const store = useEntityStore.getState();
  const all = Object.values(store.entities).filter((e) =>
    ["lock", "binary_sensor", "alarm_control_panel", "camera", "light"].includes(
      domainOf(e.entity_id),
    ),
  );
  const stepMs = 23 * 60_000; // ~one event every 23 minutes
  return all
    .map((entity, i) => ({
      when: endMs - i * stepMs,
      name: resolveName(entity, overrides),
      message: demoMessage(entity),
      entityId: entity.entity_id,
      domain: domainOf(entity.entity_id),
      state: entity.state,
    }))
    .filter((e) => e.when >= startMs && e.when <= endMs)
    .sort((a, b) => b.when - a.when);
}

// Object labels the synthetic camera events cycle through, so the demo timeline
// shows a believable mix of motion/person/vehicle markers.
const DEMO_EVENT_LABELS = ["person", "motion", "car", "motion", "dog", "person"];

/**
 * Synthesize a believable 24h spread of recorded camera events for `camera` over
 * `[startMs, endMs]`, so demo mode's timeline scrubber is populated without
 * Frigate. Events land on a steady cadence, vary their label/duration, and point
 * their thumbnail at the bundled demo poster. Returned oldest-first (timeline
 * order). Deterministic per (camera, slot) so re-fetches are stable.
 */
function synthCameraEvents(
  camera: string,
  startMs: number,
  endMs: number,
): CameraEvent[] {
  const stepMs = 37 * 60_000; // ~one event every 37 minutes
  const out: CameraEvent[] = [];
  let slot = 0;
  for (let t = startMs; t <= endMs; t += stepMs, slot++) {
    const label = DEMO_EVENT_LABELS[slot % DEMO_EVENT_LABELS.length];
    const durationMs = 20_000 + (slot % 5) * 15_000; // 20s–80s
    out.push({
      id: `demo-${camera}-${slot}`,
      camera,
      label,
      startMs: t,
      endMs: Math.min(endMs, t + durationMs),
      hasClip: true,
      hasSnapshot: true,
      thumbnailUrl: DEMO_POSTER_URL,
      snapshotUrl: DEMO_POSTER_URL,
    });
  }
  return out;
}

/** Drop one synthetic entity from the store (filtered rebuild of the map). */
function removeEntity(entityId: string): void {
  const store = useEntityStore.getState();
  const next = { ...store.entities };
  delete next[entityId];
  store.setEntities(next);
}

/** Loads the invented fixtures into the store and flags the app as demo data. */
export function createFixtureSource(): Source {
  // In-memory automation store for demo mode: configs by id, plus the synthetic
  // `automation.*` entity id we surfaced for each (so the list, toggle, and
  // "run now" behave like live HA without a real backend).
  const configs = new Map<string, AutomationConfig>();
  const entityIdFor = new Map<string, string>();

  return {
    start() {
      const map = Object.fromEntries(entities.map((e) => [e.entity_id, e]));
      const store = useEntityStore.getState();
      store.setSnapshot(map, areaRegistry);
      store.setBaseUrl(""); // demo: entity_picture paths stay page-relative
      store.setStatus("demo");
    },
    stop() {},
    async callService(domain, service, data = {}) {
      const store = useEntityStore.getState();
      const entity = data.entity_id ? store.entities[data.entity_id] : undefined;
      if (!entity) return;
      store.upsertEntities([simulate(entity, domain, service, data)]);
    },
    async fetchHistory(entityId, hours) {
      const entity = useEntityStore.getState().entities[entityId];
      if (!entity) return [];
      return synthHistory(entity, hours);
    },
    async fetchLogbook(startMs, endMs, opts) {
      const events = synthLogbook(startMs, endMs);
      if (opts?.entityIds && opts.entityIds.length > 0) {
        const want = new Set(opts.entityIds);
        return events.filter((e) => e.entityId && want.has(e.entityId));
      }
      return events;
    },
    async streamUrl(entityId) {
      // Demo "live" feed: loop the bundled clip for any camera entity.
      return domainOf(entityId) === "camera" ? DEMO_CLIP_URL : null;
    },
    async fetchCameraEvents(camera, startMs, endMs) {
      return synthCameraEvents(camera, startMs, endMs);
    },
    recordingUrlAt() {
      // Demo: every seek plays the same bundled clip (no real recordings).
      return DEMO_CLIP_URL;
    },
    eventClipUrl() {
      return DEMO_CLIP_URL;
    },
    async getAutomationConfig(id) {
      return configs.get(id) ?? null;
    },
    async saveAutomationConfig(config) {
      configs.set(config.id, config);
      // Re-surface the synthetic entity (alias may have changed → new id).
      const prev = entityIdFor.get(config.id);
      if (prev) removeEntity(prev);
      const friendlyName = String(config.alias ?? config.id);
      const entityId = `automation.${slugify(friendlyName)}`;
      entityIdFor.set(config.id, entityId);
      useEntityStore.getState().upsertEntities([
        {
          entity_id: entityId,
          state: "on",
          attributes: { friendly_name: friendlyName, id: config.id },
        },
      ]);
    },
    async deleteAutomationConfig(id) {
      configs.delete(id);
      const entityId = entityIdFor.get(id);
      if (entityId) {
        removeEntity(entityId);
        entityIdFor.delete(id);
      }
    },
  };
}

File: android/app/src/main/java/com/hawksnest/core/ha/FixtureSource.kt
package com.hawksnest.core.ha

import com.hawksnest.core.automations.slugify
import com.hawksnest.core.logic.CameraEvent
import com.hawksnest.core.logic.DEMO_CLIP_URI
import com.hawksnest.core.logic.LogEvent
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.put
import javax.inject.Inject
import kotlin.math.roundToInt

/**
 * Demo data source — no live HA. Loads the fixtures into [HaState] and simulates service calls
 * locally (the live HA source forwards them and reconciles via the state echo instead). Behind the
 * same [Source] interface as the live source, so no screen changes when we swap.
 */
class FixtureSource @Inject constructor(private val state: HaState) : Source {

    /** In-memory automation configs (id → config) + their synthetic entity ids, mirroring HA. */
    private val automationConfigs = mutableMapOf<String, JsonObject>()
    private val automationEntityIds = mutableMapOf<String, String>()

    override suspend fun start() {
        state.setBaseUrl("")
        state.setSnapshot(fixtureEntities.associateBy { it.entityId }, fixtureAreaRegistry)
        // Seed the editable configs behind the demo automation entities so the builder round-trips.
        automationConfigs.clear()
        automationConfigs.putAll(fixtureAutomationConfigs)
        automationEntityIds.clear()
        fixtureEntities.filter { domainOf(it.entityId) == "automation" }.forEach { e ->
            e.stringAttr("id")?.let { automationEntityIds[it] = e.entityId }
        }
        state.setStatus(ConnectionStatus.DEMO)
    }

    override fun stop() { /* nothing to tear down */ }

    override suspend fun callService(domain: String, service: String, data: ServiceData) {
        val id = data.entityId ?: return
        val current = state.entities.value[id] ?: return
        // Keep the current state when a service only changes attributes (set_percentage, etc.).
        val newState = simulatedState(domain, service, data) ?: current.state
        // Reflect extra service data so the demo's sliders/setpoints move.
        val attrs = current.attributes.toMutableMap()
        data.extra.forEach { (k, v) ->
            if (k == "brightness_pct") {
                (v as? Number)?.let { attrs["brightness"] = anyToJsonElement((it.toDouble() * 2.55).toInt()) }
            } else {
                attrs[k] = anyToJsonElement(v)
            }
        }
        // Mirror the live source's non-optimistic echo: we mutate the store as HA would.
        state.upsertEntities(listOf(current.copy(state = newState, attributes = JsonObject(attrs))))
    }

    /**
     * Synthesize a plausible series so the demo's entity-detail chart renders. Numeric entities get
     * a gentle deterministic wave around their current value; discrete entities hold their state.
     * The last sample matches the live state so the chart's latest point agrees with the card.
     */
    override suspend fun fetchHistory(entityId: String, hours: Int): List<HistoryPoint> {
        val current = state.entities.value[entityId] ?: return emptyList()
        val now = System.currentTimeMillis()
        val n = 24
        val stepMs = hours * 3600_000L / n
        val base = current.state.toFloatOrNull()
        return (0 until n).map { i ->
            val t = now - (n - 1 - i) * stepMs
            val s = when {
                i == n - 1 -> current.state // anchor the latest point to the live state
                base != null -> {
                    val wave = base * (1f + 0.06f * kotlin.math.sin(i.toFloat()))
                    ((wave * 10).roundToInt() / 10f).toString()
                }
                else -> current.state
            }
            HistoryPoint(t, s)
        }
    }

    /**
     * Synthesize a short demo logbook so the History tab isn't empty without a live HA — one event
     * per entity (its current state), spaced a few minutes apart and clamped to the window. Real
     * timestamps come from HA in the live source.
     */
    override suspend fun fetchLogbook(startMs: Long, endMs: Long, entityIds: List<String>?): List<LogEvent> {
        val now = System.currentTimeMillis().coerceAtMost(endMs)
        val entities = state.entities.value.values
            .filter { entityIds == null || it.entityId in entityIds }
            .take(20)
        return entities.mapIndexedNotNull { i, e ->
            val t = now - i * 7 * 60_000L
            if (t < startMs) return@mapIndexedNotNull null
            LogEvent(
                timeMs = t,
                name = e.friendlyName() ?: e.entityId,
                message = "changed to ${e.state}",
                entityId = e.entityId,
                domain = domainOf(e.entityId),
                state = e.state,
            )
        }
    }

    override suspend fun getAutomationConfig(id: String): JsonObject? = automationConfigs[id]

    /** Save in memory and upsert a synthetic `automation.<slug>` entity, as HA would after a reload. */
    override suspend fun saveAutomationConfig(config: JsonObject) {
        val id = (config["id"] as? JsonPrimitive)?.contentOrNull ?: return
        automationConfigs[id] = config
        val alias = (config["alias"] as? JsonPrimitive)?.contentOrNull ?: id
        // Reuse the existing entity id on edit so an alias change doesn't orphan the old entity.
        val entityId = automationEntityIds.getOrPut(id) { "automation.${slugify(alias)}" }
        state.upsertEntities(
            listOf(
                HassEntity(
                    entityId = entityId,
                    state = "on",
                    attributes = buildJsonObject {
                        put("friendly_name", alias)
                        put("id", id)
                    },
                ),
            ),
        )
    }

    override suspend fun deleteAutomationConfig(id: String) {
        automationConfigs.remove(id)
        val entityId = automationEntityIds.remove(id) ?: return
        state.setEntities(state.entities.value.toMutableMap().apply { remove(entityId) })
    }

    /** Demo "live" feed: loop the bundled clip for any camera entity. */
    override suspend fun streamUrl(entityId: String): String? =
        if (domainOf(entityId) == "camera") DEMO_CLIP_URI else null

    /** Demo: every seek plays the same bundled clip (no real recordings). */
    override fun recordingUrlAt(camera: String, startMs: Long, endMs: Long): String = DEMO_CLIP_URI

    override fun eventClipUrl(eventId: String): String = DEMO_CLIP_URI

    /**
     * Synthesize a believable 24h spread of recorded camera events so the demo timeline scrubber is
     * populated without Frigate. Events land on a steady cadence, vary their label/duration, and are
     * returned oldest-first. Mirrors `fixtureSource.synthCameraEvents` on the web.
     */
    override suspend fun fetchCameraEvents(camera: String, startMs: Long, endMs: Long): List<CameraEvent> {
        val stepMs = 37 * 60_000L // ~one event every 37 minutes
        val labels = listOf("person", "motion", "car", "motion", "dog", "person")
        val out = mutableListOf<CameraEvent>()
        var slot = 0
        var t = startMs
        while (t <= endMs) {
            val durationMs = 20_000L + (slot % 5) * 15_000L // 20s–80s
            out.add(
                CameraEvent(
                    id = "demo-$camera-$slot",
                    camera = camera,
                    label = labels[slot % labels.size],
                    startMs = t,
                    endMs = minOf(endMs, t + durationMs),
                    hasClip = true,
                    hasSnapshot = false,
                    thumbnailUrl = null,
                    snapshotUrl = null,
                ),
            )
            t += stepMs
            slot++
        }
        return out
    }

    /** The resulting state for a simulated `domain.service`, or null to keep the current state. */
    private fun simulatedState(domain: String, service: String, data: ServiceData): String? =
        when (service) {
            "lock" -> "locked"
            "unlock" -> "unlocked"
            "alarm_disarm" -> "disarmed"
            "alarm_arm_home" -> "armed_home"
            "alarm_arm_away" -> "armed_away"
            "alarm_arm_night" -> "armed_night"
            "open_cover" -> "open"
            "close_cover" -> "closed"
            "turn_on" -> "on"
            "turn_off" -> "off"
            "media_play_pause" -> if (state.entities.value[data.entityId]?.state == "playing") "paused" else "playing"
            else -> null
        }
}
